File: Scripts/DockerBuilds/Utilities/BuildDockerCommand.py
from CommonUtilities import run_command

class BuildDockerCommand:

    DEFAULT_BUILD_ARGUMENTS = ["base_image", "docker_image_name"]

    def __init__(self, build_arguments_keys = None):
        """
        Args:
            build_arguments_keys (list): List of build arguments keys.
        """
        if build_arguments_keys is None:
            build_arguments_keys = []
        self._build_arguments_keys = build_arguments_keys

    def run_build_docker_command(
        self,
        dockerfile_path,
        build_configuration,
        use_cache,
        build_context,
        use_host_network=False):
        """
        Builds the Docker image using the provided Dockerfile and build arguments.
        
        Args:
            dockerfile_path (Path): Path to the Dockerfile.
            build_configuration: Typically result from BuildDockerConfiguration.
            use_cache (bool): Whether to use Docker cache during build.
            build_context (Path): The directory to use as the build context.
            use_host_network (bool): Whether to use --network host during build.

        Raises:
            subprocess.CalledProcessError: If the Docker build command fails.
            ValueError: If the BASE_IMAGE is empty in the configuration.
        """
        docker_build_cmd = ["DOCKER_BUILDKIT=1", "docker", "build"]

        if not use_cache:
            docker_build_cmd.append("--no-cache")
        
        # Add network host if requested
        if use_host_network:
            docker_build_cmd.append("--network host")

        # Add build arguments
        for key in self._build_arguments_keys:
            docker_build_cmd.extend([
                "--build-arg",
                f"{key}={build_configuration[key]}"])

        # Check and add BASE_IMAGE argument
        base_image = build_configuration.base_image
        if not base_image:
            raise ValueError("BASE_IMAGE is empty in the configuration file")
        docker_image_name = build_configuration.docker_image_name

        docker_build_cmd.extend([
            "--build-arg",
            f"BASE_IMAGE={base_image}"
        ])

        # Specify Dockerfile
        docker_build_cmd.extend(["-f", str(dockerfile_path)])

        # Tag the image
        docker_build_cmd.extend(["-t", docker_image_name])

        docker_build_cmd.append(".")

        # Convert command list to string
        command_str = ' '.join(docker_build_cmd)

        run_command(command_str, cwd=build_context)

File: RustLibraries/docker_builder/src/build_docker/build_docker_command.rs
use std::collections::HashMap;
use std::path::Path;

use crate::configuration::build_docker_configuration::BuildDockerConfigurationData;

/// Builds the Docker build command as a list of arguments.
///
/// # Arguments
/// * `dockerfile_path` - Path to the Dockerfile
/// * `build_configuration` - Populated BuildDockerConfigurationData
/// * `use_cache` - Whether to use Docker build cache
/// * `use_host_network` - Whether to use --network host
///
/// # Returns
/// * Vec<String> representing the full command (e.g., for exec or logging)
///
/// Note: Build context (.) and cwd are handled separately at runtime if needed.
pub fn build_docker_build_command(
    dockerfile_path: &Path,
    build_configuration: &BuildDockerConfigurationData,
    use_cache: bool,
    use_host_network: bool,
) -> Vec<String> {
    let mut docker_build_cmd = vec![
        "DOCKER_BUILDKIT=1".to_string(),
        "docker".to_string(),
        "build".to_string(),
    ];

    if !use_cache {
        docker_build_cmd.push("--no-cache".to_string());
    }

    if use_host_network {
        docker_build_cmd.push("--network".to_string());
        docker_build_cmd.push("host".to_string());
    }

    // Add build_args from configuration (dynamic from YAML)
    let build_args: &HashMap<String, String> = &build_configuration.build_args;
    for (key, value) in build_args {
        docker_build_cmd.push("--build-arg".to_string());
        docker_build_cmd.push(format!("{}={}", key.to_uppercase(), value));
    }

    // Always add base_image and docker_image_name as --build-arg
    if !build_configuration.base_image.is_empty() {
        docker_build_cmd.push("--build-arg".to_string());
        docker_build_cmd.push(format!("BASE_IMAGE={}", build_configuration.base_image));
    } else {
        // In real use, this could be logged/handled upstream
        eprintln!("Warning: base_image is empty in configuration");
    }

    if !build_configuration.docker_image_name.is_empty() {
        docker_build_cmd.push("--build-arg".to_string());
        docker_build_cmd.push(format!("DOCKER_IMAGE_NAME={}", build_configuration.docker_image_name));
    } else {
        eprintln!("Warning: docker_image_name is empty in configuration");
    }

    // Specify Dockerfile
    docker_build_cmd.push("-f".to_string());
    docker_build_cmd.push(dockerfile_path.to_string_lossy().into_owned());

    // Tag the image
    docker_build_cmd.push("-t".to_string());
    docker_build_cmd.push(build_configuration.docker_image_name.clone());

    // Build context (.)
    docker_build_cmd.push(".".to_string());

    docker_build_cmd
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    #[test]
    fn test_build_docker_build_command() {
        let mut build_args = HashMap::new();
        build_args.insert("ARG1".to_string(), "value1".to_string());

        let config = BuildDockerConfigurationData {
            docker_image_name: "test-image:latest".to_string(),
            base_image: "ubuntu:22.04".to_string(),
            build_args,
            dockerfile_components: vec![],
        };

        let dockerfile_path = Path::new("Dockerfile");
        let cmd = build_docker_build_command(
            dockerfile_path,
            &config,
            false,
            true,
        );

        // Uncomment the following and run `cargo test -- --nocapture` to
        // inspect the full command during tests:
        // println!("Built Docker command: {:?}", cmd);

        assert!(cmd.contains(&"--no-cache".to_string()));
        assert!(cmd.contains(&"--network".to_string()));
        assert!(cmd.contains(&"host".to_string()));
        assert!(cmd.contains(&"--build-arg".to_string()));
        assert!(cmd.iter().any(|s| s.contains("ARG1=value1")));
        assert!(cmd.iter().any(|s| s.contains("BASE_IMAGE=ubuntu:22.04")));
        assert!(cmd.iter().any(|s| s.contains(
            "DOCKER_IMAGE_NAME=test-image:latest")));
        assert!(cmd.last() == Some(&".".to_string()));
    }
}
